### PR TITLE
Add the option for particle RF phase-wrapping

### PR DESCRIPTION
**Pull request by cemitch99**
This PR adds the capability of applying particle boundary conditions, related to the RF bucket (for example, wrapping particles modulo 2pi in RF phase).

- [x] add bucket length to particle container
- [x] add user-facing input options for bucket length
- [x] add function applying periodic particle boundary conditions
- [x] add functionality for alternative boundary options (e.g. reflection)
- [x] add Python bindings
- [x] add user-facing example (periodic bc)
- [x] add Python unit tests
- [x] update documentation
- [x] test absorbing & reflecting
- [x] finish reflecting or document approximation

Longitudinal particle boundary condition options to target:

1. open/off (no action)
2. periodic (phase wrapping)
3. cut (particle lost)
4. reflection (particle bounces back with reversed momentum)

### CODE REVIEW

File: src/particles/ParticleBoundary.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+                amrex::ParticleReal* const AMREX_RESTRICT part_t = soa_real[RealSoA::t].dataPtr();
+
+                // Gather particles and apply boundary condition
+                amrex::ParallelFor(np, [=] AMREX_GPU_DEVICE (int i)
+                {
+                    // Access SoA Real data
+                    amrex::ParticleReal & AMREX_RESTRICT t = part_t[i];
+
+                    if(particle_bc == "periodic") {
```
Comment (cemitch99) [inline src/particles/ParticleBoundary.cpp]: This conditional is causing the build tests to fail.  Moving this conditional outside the loop(s) fixes the issue, but requires duplication of the loop within each branch.  There should be a cleaner solution (one that avoids duplication).

File: examples/phase_wrapper/input_drift_phase_wrap.in [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+beam.muxpx = -0.846574929020762
+beam.muypy = 0.846574929020762
+beam.mutpt = 0.0
+beam.bucket_length = 0.23
+
+###############################################################################
+# Beamline: lattice elements and segments
+###############################################################################
+lattice.periods = 10
+#lattice.elements = monitor drift1 monitor
```
Comment (ax3l) [inline examples/phase_wrapper/input_drift_phase_wrap.in]: Dead code:
```suggestion
```

File: src/particles/ImpactXParticleContainer.H [hunk 3]
```diff
@@ -305,6 +305,19 @@ namespace impactx
...
 
+        /** Set the bucket length used for periodic particle BCs in this container
+         *
+         * @param bucket_length the length of the RF bucket in m
+         */
+        void SetBucketLength (amrex::ParticleReal bucket_length);
+
+        /** Get the bucket length used for periodic particle BCs in this container
+         *
+         * @returns bucket_length
```
Comment (ax3l) [inline src/particles/ImpactXParticleContainer.H]:
```suggestion
         * @returns bucket length in m
```

File: src/particles/ImpactXParticleContainer.H [hunk 4]
```diff
@@ -305,6 +305,19 @@ namespace impactx
...
+         */
+        amrex::ParticleReal &
```
Comment (ax3l) [inline src/particles/ImpactXParticleContainer.H]:
```suggestion
        amrex::ParticleReal
```

File: src/particles/ImpactXParticleContainer.cpp [hunk 5]
```diff
@@ -123,6 +123,17 @@ namespace impactx
         SetParticleShape(v);
     }
 
+    void ImpactXParticleContainer::SetBucketLength (amrex::ParticleReal bucket_length)
+    {
+        m_bucket_length = bucket_length;
+    }
+
+    amrex::ParticleReal &
```
Comment (ax3l) [inline src/particles/ImpactXParticleContainer.cpp]:
```suggestion
    amrex::ParticleReal
```

File: src/particles/ImpactXParticleContainer.H [hunk 6]
```diff
@@ -341,6 +354,9 @@ namespace impactx
         //! history of the beam moments over s
         std::list<std::unordered_map<std::string, amrex::ParticleReal> > m_beam_moments;
 
+        //! the RF bucket length for particles in the particle container
+        amrex::ParticleReal m_bucket_length;
```
Comment (ax3l) [inline src/particles/ImpactXParticleContainer.H:358]: I would store this as `std::optional` so it is clear when this is thought to be undefined/None:
```suggestion
        std::optional<amrex::ParticleReal> m_bucket_length;
```

Maybe even store like `store_beam_moments` as a public member and document that instead.

File: src/particles/ParticleBoundary.cpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    )
+    {
+        BL_PROFILE("impactx::particles::ParticleBoundary")
+
+        using namespace amrex::literals;
+
+        amrex::ParmParse pp_algo("algo");
+        std::string particle_bc = "open";
+        pp_algo.queryAdd("particle_bc", particle_bc);
+        int particle_bc_int;
```
Comment (ax3l) [inline src/particles/ParticleBoundary.cpp]: Good case for an `enum`, in our case we can use `AMREX_ENUM` that supports string conversion.

File: src/particles/ParticleBoundary.cpp [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        if(particle_bc == "open") {
+            return;
+        } else if (particle_bc == "periodic") {
+            particle_bc_int = 1;
+        } else if (particle_bc == "absorbing") {
+            particle_bc_int = 2;
+        } else if (particle_bc == "reflecting") {
+            particle_bc_int = 3;
+        } else {
+            particle_bc_int = 0;
```
Comment (ax3l) [inline src/particles/ParticleBoundary.cpp]: Logic bug: else seems to imply `open`, but `open` already returned above.

Usability: we should throw an error if an unknown other BC is set than the ones we know/handle. (This does not prevent handling no input as _open_ by default.)

File: src/particles/ParticleBoundary.cpp [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+                {
+                    // Access SoA Real data
+                    amrex::ParticleReal & AMREX_RESTRICT t = part_t[i];
+                    amrex::ParticleReal & AMREX_RESTRICT pt = part_pt[i];
+
+                    if (particle_bc_int==1) {
+
+                        // Periodic particle boundary condition:  apply phase wrapping in t (modulo bucket_duration):
+                        amrex::ParticleReal ttest = std::fmod(t+bucket_half_duration, bucket_duration);
+                        t = (bucket_duration != 0.0)? std::fmod(ttest+bucket_duration, bucket_duration)-bucket_half_duration : t;
```
Comment (ax3l) [inline src/particles/ParticleBoundary.cpp]: Suggestion: we can write the `if`/`switch` on the enum `particle_bc_int` outside the `amrex::ParallelFor` and create 3 unique, performant `ParallelFor` kernels; instead of one branching (slow) one.

File: src/particles/ParticleBoundary.H [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "particles/ImpactXParticleContainer.H"
+
+#include <AMReX_GpuContainers.H>
+#include <AMReX_REAL.H>
+
+
+namespace impactx::particles
+{
+    /** Function to apply longitudinal particle boundary (periodic, etc.)
```
Comment (ax3l) [inline src/particles/ParticleBoundary.H:21]: Add the `AMREX_ENUM` on the types of boundaries we have here.

File: src/particles/ParticleBoundary.cpp [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    void ParticleBoundary (
+        ImpactXParticleContainer & pc
+    )
+    {
+        BL_PROFILE("impactx::particles::ParticleBoundary")
+
+        using namespace amrex::literals;
+
+        amrex::ParmParse pp_algo("algo");
+        std::string particle_bc = "open";
```
Comment (ax3l) [inline src/particles/ParticleBoundary.cpp:28]: This feel like something we could set on `beam.[...]` was well... (probably need to [generalize the section heading here](https://impactx.readthedocs.io/en/latest/usage/parameters.html#initial-beam-distributions))

But no super strong opinion.

File: src/particles/ParticleBoundary.cpp (hunk 7 above)
Comment (cemitch99) [reply src/particles/ParticleBoundary.cpp]: Yes, thanks; I will revisit here.

File: src/particles/ParticleBoundary.cpp (hunk 8 above)
Comment (cemitch99) [reply src/particles/ParticleBoundary.cpp]: Agreed; will replace this line with an error message.

File: src/particles/ParticleBoundary.cpp (hunk 9 above)
Comment (cemitch99) [reply src/particles/ParticleBoundary.cpp]: Oh, I thought about branching outside the `ParallelFor`, but it introduces a lot of duplication.  It's ok with me to rework things this way if there is an expected performance advantage.

File: src/particles/ImpactXParticleContainer.cpp [hunk 12]
```diff
@@ -123,6 +123,17 @@ namespace impactx
...
 
+    void ImpactXParticleContainer::SetBucketLength (amrex::ParticleReal bucket_length)
+    {
+        m_bucket_length = bucket_length;
+    }
+
+    amrex::ParticleReal
+    ImpactXParticleContainer::GetBucketLength ()
+    {
+        return m_bucket_length;
```
Comment (cemitch99) [inline src/particles/ImpactXParticleContainer.cpp:134]:
```suggestion
        if (m_bucket_length.has_value()) {
            return m_bucket_length;
        } else {
            return 0.0;
        }
```

File: src/particles/ImpactXParticleContainer.cpp [hunk 13]
```diff
@@ -123,6 +123,21 @@ namespace impactx
...
+    }
+
+    amrex::ParticleReal
+    ImpactXParticleContainer::GetBucketLength ()
+    {
+        if (m_bucket_length.has_value()) {
+            return m_bucket_length;
+        } else {
+            return 0.0;
+        }
```
Comment (cemitch99) [inline src/particles/ImpactXParticleContainer.cpp]:
```suggestion
        return (m_bucket_length.has_value()?) m_bucket_length : 0.0;
```

File: src/particles/ImpactXParticleContainer.H [hunk 14]
```diff
@@ -341,6 +354,9 @@ namespace impactx
         //! history of the beam moments over s
         std::list<std::unordered_map<std::string, amrex::ParticleReal> > m_beam_moments;
 
+        //! the RF bucket length for particles in the particle container
+        std::optional<amrex::ParticleReal> m_bucket_length;
```
Comment (cemitch99) [inline src/particles/ImpactXParticleContainer.H]: @ax3l Making this optional is introducing build problems that are taking time to fix.  Reverting to the previous behavior for now.

Comment (cemitch99) [inline src/particles/ImpactXParticleContainer.H]:
```suggestion
        amrex::ParticleReal m_bucket_length;
```

File: src/particles/ImpactXParticleContainer.cpp [hunk 15]
```diff
@@ -123,6 +123,17 @@ namespace impactx
...
 
+    void ImpactXParticleContainer::SetBucketLength (amrex::ParticleReal bucket_length)
+    {
+        m_bucket_length = bucket_length;
+    }
+
+    amrex::ParticleReal
+    ImpactXParticleContainer::GetBucketLength ()
+    {
+        return (m_bucket_length.has_value()?) m_bucket_length : 0.0;
```
Comment (cemitch99) [inline src/particles/ImpactXParticleContainer.cpp]:
```suggestion
        return m_bucket_length;
```

File: src/particles/ImpactXParticleContainer.H (hunk 14 above)
Comment (ax3l) [reply src/particles/ImpactXParticleContainer.H]: Oh yes, I just drafted the implementation. I can do it when the PR is ready from your end.

File: examples/phase_wrapper/run_drift_phase_wrap.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+sim.add_particles(bunch_charge_C, distr, npart)
+
+# add beam diagnostics
+monitor = elements.BeamMonitor("monitor", backend="h5")
+
+# design the accelerator lattice)
+ns = 1  # number of slices per ds in the element
+line = [
+    monitor,
+    elements.Drift(name="drift1", ds=0.0, nslice=ns),
```
Comment (ax3l) [inline examples/phase_wrapper/run_drift_phase_wrap.py]: Could be worth testing:
```suggestion
```

Warp is applied per element slice and would be done at the end of the first monitor.

File: src/particles/ParticleBoundary.cpp [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+        pp_algo.queryAdd("particle_bc", particle_bc);
+
+        int particle_bc_int = 0;
+
+        if (particle_bc == "periodic") {
+            particle_bc_int = 1;
+        } else if (particle_bc == "absorbing") {
```
Comment (ax3l) [inline src/particles/ParticleBoundary.cpp:35]: memo for @ax3l : use `AMREX_ENUM` (define in `ParticleBoundary.H`)

File: src/particles/ImpactXParticleContainer.H (hunk 6 above)
Comment (ax3l) [reply src/particles/ImpactXParticleContainer.H:358]: memo to myself: do at the end

File: src/particles/ParticleBoundary.cpp (hunk 17 above)
Comment (ax3l) [reply src/particles/ParticleBoundary.cpp]: added

File: src/particles/ParticleBoundary.cpp [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+#endif
+            for (ParIt pti(pc, lev); pti.isValid(); ++pti)
+            {
+                const int np = pti.numParticles();
+
+                // Access data from StructOfArrays (soa)
+                auto & soa_real = pti.GetStructOfArrays().GetRealData();
+
+                amrex::ParticleReal * const AMREX_RESTRICT part_t = soa_real[RealSoA::t].dataPtr();
+                amrex::ParticleReal * const AMREX_RESTRICT part_pt = soa_real[RealSoA::t].dataPtr();
```
Comment (ax3l) [inline src/particles/ParticleBoundary.cpp]: Bug:

```suggestion
                amrex::ParticleReal * const AMREX_RESTRICT part_pt = soa_real[RealSoA::pt].dataPtr();
```

Comment (ax3l) [reply src/particles/ParticleBoundary.cpp]: fixed.

File: src/particles/ParticleBoundary.cpp (hunk 9 above)
Comment (ax3l) [reply src/particles/ParticleBoundary.cpp]: Thanks, completed :)

File: src/particles/ParticleBoundary.H (hunk 10 above)
Comment (ax3l) [reply src/particles/ParticleBoundary.H:32]: done

File: src/particles/ParticleBoundary.cpp (hunk 11 above)
Comment (ax3l) [reply src/particles/ParticleBoundary.cpp]: we kept bucket len on beam and bc on algo for now (physics / algorithm-ish)

File: src/particles/ImpactXParticleContainer.H (hunk 14 above)
Comment (ax3l) [reply src/particles/ImpactXParticleContainer.H]: I decided against it in the end, too

File: src/particles/ParticleBoundary.cpp [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+                            amrex::ParticleReal & AMREX_RESTRICT t = part_t[i];
+                            amrex::ParticleReal & AMREX_RESTRICT pt = part_pt[i];
+
+                            // Reflecting particle boundary condition.
+                            // TODO:  Transform (t,pt) to (z,pz) using z-to-t transformation.
+                            // The implementation below works through linear order in the phase space variables.
+                            // If particle falls outside the boundary, reflect:
+                            if (t > bucket_half_duration) {
+                                t = bucket_duration - t;
+                                pt = -pt;
```
Comment (ax3l) [inline src/particles/ParticleBoundary.cpp:124]: If you like we can fix that by adding the transforms, either in-line in the parallel for or outside the level loop.

File: docs/source/usage/python.rst [hunk 20]
```diff
@@ -203,6 +203,21 @@ Collective Effects & Overall Simulation Parameters
...
+
+      Options:
+
+      * ``"open"`` (default): no action is applied at the boundary.
+
+      * ``"periodic"``: each particle's longitudinal coordinate is treated modulo the value ``bucket_length`` (phase wrapping).
+
+      * ``"absorbing"``: a particle whose longitudinal coordinate falls outside the boundary is declared lost.
+
+      * ``"reflecting"``: a particle whose longitudinal coordinate crosses the boundary is reflected about the boundary, with reversed longitudinal momentum.
```
Comment (ax3l) [inline docs/source/usage/python.rst:219]: Approximation not mentioned: pls see comment in summary.

File: docs/source/usage/parameters.rst [hunk 21]
```diff
@@ -1191,10 +1194,32 @@ However, a Taylor expansion is used to evaluate the dependence on the quantum pa
...
+
+  The type of particle boundary condition to be applied to the longitudinal coordinate, based on the value of parameter ``bucket_length``.  Options:
+
+  * ``"open"`` (default): no action is applied at the boundary.
+
+  * ``"periodic"``: each particle's longitudinal coordinate is treated modulo the value ``bucket_length`` (phase wrapping).
+
+  * ``"absorbing"``: a particle whose longitudinal coordinate falls outside the boundary is declared lost.
+
+  * ``"reflecting"``: a particle whose longitudinal coordinate crosses the boundary is reflected about the boundary, with reversed longitudinal momentum.
```
Comment (ax3l) [inline docs/source/usage/parameters.rst:1216]: Approximation not yet mentioned here.

**Review by ax3l — COMMENTED** [review]
The implementation and interfaces look great to me now!

The cases `absorbing` and `reflecting` are still uncovered in tests, the latter had a copy-paste bug that a test would have caught. If you like, we can add small `pytests` for them in case examples are too verbose.

The `reflecting` logic is an approximation that needs to either be documented in `parameters.rst` + `python.rst` or we add the transforms.

File: tests/python/test_particle_bc.py [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+
+    # numerical parameters
+    bucket_length = 0.2
+    half_bucket_length = bucket_length / 2.0
+
+    # check that all final particle lie within the bucket
+    np.testing.assert_array_less([zmax], [half_bucket_length])
+
+
+def test_particle_bc_reflecting():
```
Comment (ax3l) [inline tests/python/test_particle_bc.py:93]: Memo to myself: reduplicate common part of those 3 tests

**Review by ax3l — APPROVED** [review]
Thank you, @cemitch99 ! 🎉

File: src/particles/ParticleBoundary.cpp [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+    {
+        BL_PROFILE("impactx::particles::ParticleBoundary")
+
+        // check option and set default (open) if unset
+        auto particle_bc = get_particle_boundary_condition();
+
+        // nothing to do by default
+        if (particle_bc == ParticleBC::open) {
+            return;
+        }
```
Comment (ax3l) [inline src/particles/ParticleBoundary.cpp]: That way it does not show up in the profiler if unused:
```suggestion
        // check option and set default (open) if unset
        auto particle_bc = get_particle_boundary_condition();

        // nothing to do by default
        if (particle_bc == ParticleBC::open) {
            return;
        }

        BL_PROFILE("impactx::particles::ParticleBoundary")
```

File: docs/source/usage/python.rst (hunk 20 above)
Comment (ax3l) [reply docs/source/usage/python.rst:219]: done

File: docs/source/usage/parameters.rst (hunk 21 above)
Comment (ax3l) [reply docs/source/usage/parameters.rst:1216]: done

File: tests/python/test_particle_bc.py (hunk 22 above)
Comment (ax3l) [reply tests/python/test_particle_bc.py:93]: done
